subprojects/platform-native/src/main/java/org/gradle/nativeplatform/toolchain/internal/msvcpp/AbstractWindowsKitComponentLocator.java

```diff
@@ -0,0 +1,282 @@
+        "SOFTWARE\\",
+        "SOFTWARE\\Wow6432Node\\"
+    };
+    private static final String REGISTRY_ROOTPATH_KIT = "Microsoft\\Windows Kits\\Installed Roots";
+    private static final String REGISTRY_KIT_10 = "KitsRoot10";
+
+    private final Pattern windowsKitVersionPattern = Pattern.compile("[0-9]+(\\.[0-9]+)*");
+    private final FileFilter windowsKitVersionFilter = new FileFilter() {
+        @Override
+        public boolean accept(File pathname) {
+            Matcher matcher = windowsKitVersionPattern.matcher(pathname.getName());
+            return pathname.isDirectory() && matcher.matches();
+        }
+    };
+
+    AbstractWindowsKitComponentLocator(WindowsRegistry windowsRegistry) {
+        this.windowsRegistry = windowsRegistry;
+    }
+
+    @Override
+    public WindowsKitComponentLocator.SearchResult<T> locateComponents(File candidate) {
+        initializeComponents();
+
+        if (candidate != null) {
+            return locateUserSpecifiedComponent(candidate);
+        }
+
+        return locateDefaultComponent();
+    }
+
+    @Override
+    public List<T> locateAllComponents() {
+        initializeComponents();
+
+        List<T> allComponents = Lists.newArrayList();
+        for (Set<T> components : foundComponents.values()) {
+            allComponents.addAll(components);
+        }
+        return allComponents;
+    }
+
+    private void initializeComponents() {
+        if (!initialised) {
+            locateComponentsInRegistry();
+            initialised = true;
+        }
+    }
+
+    private WindowsKitComponentLocator.SearchResult<T> locateDefaultComponent() {
+        T selected = getBestComponent();
+        return selected == null
+            ? new ComponentNotFound("Could not locate a " + getDisplayName() + " installation using the Windows registry.")
+            : new ComponentFound(selected);
+    }
+
+    private T getBestComponent(File baseDir) {
+        final SortedSet<T> candidates = new TreeSet<T>(new DescendingComponentVersionComparator());
+        Set<T> components = foundComponents.get(baseDir);
+        if (components != null) {
+            candidates.addAll(components);
+        }
+        return candidates.isEmpty() ? null : candidates.iterator().next();
+    }
+
+    private T getBestComponent() {
+        final SortedSet<T> candidates = new TreeSet<T>(new DescendingComponentVersionComparator());
+        for (Set<T> components : foundComponents.values()) {
+            candidates.addAll(components);
+        }
+        return candidates.isEmpty() ? null : candidates.iterator().next();
+    }
+
+    private void locateComponentsInRegistry() {
+        for (String baseKey : REGISTRY_BASEPATHS) {
+            locateComponentsInRegistry(baseKey);
+        }
+    }
+
+    private void locateComponentsInRegistry(String baseKey) {
+        String[] keys = {
+                REGISTRY_KIT_10
+        };
+
+        for (String key : keys) {
```

Comment:
This is iterating over an array with a single element in it. Looks like it was already here, but could you clean this up while we're in here?

Reply:
Yeah, that seems a little unnecessary.  Removed.